jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAccumulator.java

```diff
@@ -105,6 +120,94 @@ public void copyBuffer(ByteBuffer source)
         }
     }
 
+    public void addBuffer(RetainableByteBuffer buffer)
+    {
+        if (buffer != null && buffer.hasRemaining())
+        {
+            if (maxLengthExceeded(buffer.remaining()))
+                throw new IllegalArgumentException("maxLength exceeded");
+
+            if (buffer.canRetain())
+            {
+                buffer.retain();
+                _buffers.add(buffer);
+            }
+            else
+            {
+                copyBuffer(buffer.getByteBuffer());
+            }
+        }
+    }
+
+    /**
+     * Aggregates the given ByteBuffer into the last buffer of this accumulation, growing the buffer in size if
+     * necessary. This copies bytes up to the specified maximum size into the
+     * last buffer in the accumulation, at which time this method returns {@code true}
+     * and {@link #takeRetainableByteBuffer()} must be called for this method to accept aggregating again.
+     * @param source the buffer to copy into this aggregator; its position is updated according to
+     * the number of aggregated bytes
+     * @return true if the aggregator's buffer is full and should be taken, false otherwise
+     */
+    public boolean aggregate(ByteBuffer source)
+    {
+        if (BufferUtil.isEmpty(source))
+            return false;
+
+        // How much of the buffer can be aggregated?
+        int toCopy = source.remaining();
+        boolean full = false;
+        if (_maxLength >= 0)
+        {
+            long space = _maxLength - getLength();
+            if (space == 0)
+                return true;
+            if (toCopy >= space)
+            {
+                full = true;
+                toCopy = (int)space;
+            }
+        }
+
+        // Do we need to allocate a new buffer?
+        RetainableByteBuffer buffer = _buffers.isEmpty() ? null : _buffers.get(_buffers.size() - 1);
+        if (buffer == null || buffer.isRetained() || BufferUtil.space(buffer.getByteBuffer()) < toCopy)
+        {
+            int prefix = buffer == null ? 0 : buffer.remaining();
+            int minSize = prefix + toCopy;
+            int allocSize = (int)Math.min(_maxLength, ceilToNextPowerOfTwo(minSize));
```

Comment:
Should this be min of the remaining space `_maxLength - getLength()` instead of just `_maxLength`?

Reply:
@lachlan-roberts I think it already is, since the `toCopy` has already been limited by `space = _maxLength - getLength()` above.